File: js/1496.判断路径是否相交.js
/**
 * @param {string} path
 * @return {boolean}
 */
var isPathCrossing = function(path) {
    var mapping = {}, x = 0, y = 0;
    mapping[x.toString() + y.toString()] = true;
    for (var i = 0; i < path.length; i ++) {
        if (path[i] === 'N') y ++;
        else if (path[i] === 'S') y --;
        else if (path[i] === 'W') x --;
        else x ++;

        let str = x.toString() + y.toString();
        if (mapping[str] === undefined) mapping[str] = true;
        else return true; 
    }
    
    return false;
};

File: rust/1496.判断路径是否相交.rs
impl Solution {
    pub fn is_path_crossing(path: String) -> bool {
        use std::collections::HashSet;

        let (mut x, mut y) = (0, 0);
        let mut hs = HashSet::new();
        hs.insert((x, y));
        for char in path.chars().into_iter() {
            if char == 'N' {
                y += 1;
            }
            else if char == 'S' {
                y -= 1;
            }
            else if char == 'E' {
                x += 1;
            }
            else {
                x -= 1;
            }
            if hs.contains(&(x, y)) {
                return true;
            }
            else {
                hs.insert((x, y));
            }
        }
        false
    }
}
